File: src/main/java/org/example/sorting/BogoSort.java
package org.example.sorting;

import java.util.Random;

public class BogoSort implements org.example.SortingAlgorithm {
    @Override
    public void sort(int[] arr) {
        while (!isSorted(arr)) {
            shuffle(arr);
        }
    }

    private boolean isSorted(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                return false;
            }
        }
        return true;
    }

    static void shuffle(int[] arr) {
        Random random = new Random();
        for (int i = 0; i < arr.length; i++) {
            int randomIndex = random.nextInt(arr.length);
            int temp = arr[i];
            arr[i] = arr[randomIndex];
            arr[randomIndex] = temp;
        }
    }
}

File: algorithms/bogo_sort.py
# bogo_sort.py
import random


def is_sorted(arr):
    return all(arr[i] <= arr[i + 1] for i in range(len(arr) - 1))


def bogo_sort(arr):
    """
    Bogo Sort – randomly shuffles until sorted. (Educational only.)

    Time Complexity: Average-case O((n+1)!)
    Space Complexity: O(1)
    """
    if not arr:
        return []
    a = arr[:]
    while not is_sorted(a):
        random.shuffle(a)
    return a
